asf_heat_pump_suitability/pipeline/prepare_features/anchor_properties.py

```diff
@@ -0,0 +1,170 @@
+    """
+    logger.info("Loading POI data...")
+    poi = load_s3_data("asf-heat-pump-suitability", "source_data/poi_uk.gpkg")
+
+    # Validate POI data
+    required_columns = {"country", "main_category", "geometry"}
+    missing_columns = required_columns - set(poi.columns)
+    if missing_columns:
+        raise ValueError(f"Missing required columns in POI data: {missing_columns}")
+
+    # Filter and process
+    poi = poi[poi.country == '"GB"'].copy()
+    poi["main_category"] = poi["main_category"].apply(
+        lambda x: str.strip(x, '"') if pd.notna(x) else None
+    )
+
+    # Filter anchor properties and reproject
+    anchor_properties = (
+        poi[poi.main_category.isin(ANCHOR_PROPERTIES)]
+        .to_crs(INPUT_CRS)
+        .to_crs(PROCESSING_CRS)
+    )
+
+    logger.info(f"Found {len(anchor_properties)} potential anchor properties")
+    return anchor_properties
+
+
+def identify_anchor_properties() -> gpd.GeoDataFrame:
+    """
+    Identify and analyze anchor properties within LSOAs.
+
+    Returns:
+        GeoDataFrame: Summary of anchor properties by LSOA containing columns:
+            - lsoa: Unique identifier for the LSOA
+            - lsoa_name: Name of the LSOA
+            - anchor_count: Number of anchor properties in the LSOA
+            - building_categories: List of main categories present
+            - building_subcategories: List of subcategories present
+            - has_anchor_property: Boolean indicating presence of anchor properties
+    """
+    try:
+        logger.info("Starting anchor property analysis...")
+
+        # Load required data
+        lsoa_gdf = logger.info("Loading LSOA boundaries...")
+        lsoa_gdf = gpd.read_file(
+            "s3://asf-heat-pump-suitability/source_data/Lower_layer_Super_Output_Areas_2021_EW_BFE_V9_-9107090204806789093/LSOA_2021_EW_BFE_V9.shp"
+        )
+        anchor_properties = load_and_process_poi()
+
+        lsoa_with_anchors = gpd.sjoin(
+            lsoa_gdf, anchor_properties, how="left", predicate="intersects"
+        )
+
+        lsoa_anchor_summary = (
+            lsoa_with_anchors.groupby("LSOA21CD_left")
+            .agg(
+                {
+                    "LSOA21NM_left": "first",
+                    "id": "count",
+                    "main_category": _safe_join_categories,
+                    "alternate_category": _safe_join_categories,
+                }
+            )
+            .reset_index()
+        )
+
+        lsoa_anchor_summary.columns = [
+            "lsoa",
+            "lsoa_name",
+            "anchor_count",
+            "building_categories",
+            "building_subcategories",
+        ]
+
+        lsoa_anchor_summary["has_anchor_property"] = (
+            lsoa_anchor_summary["anchor_count"] > 0
+        )
+
+        logger.info(
+            f"Found {lsoa_anchor_summary['has_anchor_property'].sum()} LSOAs with suitable anchor properties"
+        )
+
+        return lsoa_anchor_summary
```

Comment:
There's some LSOAs that have a really high number of anchor properties. 418 have more than 20 which is high but I guess believable in a dense city centre but might need a bit further investigation and checks if you haven't done so already. E.g. I can see one LSOA has 184. I think this LSOA contains Heathrow Airport. But I wonder if some POIs are kind of getting double counted, like should we be counting shops in the airport as separate properties? Or for example, if you have a university&NHS hospital, could it be tagged multiple times in the dataset as 'hospital', 'public_service_and_government' and 'university_college' ?

If so, maybe we need to add some deduplication into our processing of the anchor properties data, either of duplicate point geoms or ones which are within a certain threshold distance of each other.

Reply:
I agree that such high numbers of anchor properties sounds unrealistic. I think deduplication based on point geoms is feasible but setting a threshold based on distance to each other from a point would be quite arbitrary since we don't know how much area around that point the building actually covers. For example, there would be quite a big difference between the optimal threshold for an airport compared to a hospital.

I think the best solution to this issue would be to incorporate building footprint data into this feature which could be used to deduplicate by checking overlaps or distance between outer bounds.

Reply:
Sounds good, thanks for checking this.

I think the pipeline is missing deduplication of the POIs? I don't see it.